New Admin UI - Elevation, Icon, and Card Components

adrians5j opened the pull request:
This PR creates the Card component.

Had to pull some fixes from [here](https://github.com/shadcn-ui/ui/pull/2439/files), and now it looks good even when there's no header/footer provided.

![image](https://github.com/user-attachments/assets/449bda79-9b90-4658-b942-504c86bd4418)

adrians5j committed:
wip: create button/circular progress components

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/new-admin-ui' into feat/button

adrians5j committed:
wip: create button/circular progress components

adrians5j committed:
wip: create circular progress component

adrians5j committed:
wip: create circular progress component

adrians5j committed:
wip: create circular progress component

adrians5j committed:
wip: create circular progress component

adrians5j committed:
wip: create circular progress component

adrians5j committed:
wip: create circular progress component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create circular progress component

adrians5j committed:
ci: skip Jest tests for PRs targeting new-admin-ui branch

adrians5j committed:
fix: add component in admin-ui

adrians5j committed:
wip: create button component

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/circular-progress' into feat/button

adrians5j committed:
Merge remote-tracking branch 'origin/feat/button' into feat/button

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
fix: move to admin-ui package (ui-new)

adrians5j committed:
fix: undo testing-related changes

adrians5j committed:
fix: move to admin-ui package (ui-new)

adrians5j committed:
fix: move to admin-ui package (ui-new)

adrians5j committed:
Merge branch 'refs/heads/adrian/switch-to-admin-ui' into feat/circular-progress

adrians5j committed:
fix: move to admin-ui package (ui-new)

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/new-admin-ui' into adrian/switch-to-admin-ui

adrians5j committed:
fix: update dependencies

adrians5j committed:
Merge branch 'refs/heads/adrian/switch-to-admin-ui' into feat/circular-progress

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/circular-progress' into feat/button

adrians5j committed:
fix: update dependencies

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/new-admin-ui' into feat/circular-progress

adrians5j committed:
fix: update dependencies

adrians5j committed:
Merge branch 'refs/heads/feat/circular-progress' into feat/button

adrians5j committed:
fix: update dependencies

adrians5j committed:
Merge branch 'refs/heads/feat/new-admin-ui' into feat/circular-progress

adrians5j committed:
Merge branch 'refs/heads/feat/circular-progress' into feat/button

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create CP component

adrians5j committed:
Merge branch 'refs/heads/feat/circular-progress' into feat/button

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/new-admin-ui' into feat/button

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create button component

adrians5j committed:
wip: create card component

adrians5j committed:
fix: use Icon component

adrians5j committed:
fix: add prop descriptions

adrians5j committed:
fix: export a component that's marked as deprecated

adrians5j committed:
fix: add deprecation notice

adrians5j committed:
fix: add deprecation notice

adrians5j committed:
fix: refactor / cleanup components

adrians5j committed:
fix: refactor / cleanup components

adrians5j committed:
chore: run prettier

adrians5j committed:
Merge branch 'refs/heads/feat/button' into feat/card

adrians5j committed:
wip: create card component

adrians5j committed:
fix: use `ghost` variant [no ci]

adrians5j committed:
Merge branch 'refs/heads/feat/button' into feat/card

adrians5j committed:
wip: create Card component

adrians5j committed:
wip: create Card component

adrians5j committed:
wip: create Card component

adrians5j committed:
fix: add `iconPosition` prop

adrians5j committed:
fix: change icon color for primary button

adrians5j committed:
Merge branch 'refs/heads/feat/button' into feat/card

adrians5j committed:
fix: export with `type` keyword

adrians5j committed:
chore: run prettier

leopuleo commented on packages/admin-ui/src/Card/Card.tsx:
```diff
@@ -1,56 +1,104 @@
+    HTMLParagraphElement,
+    React.HTMLAttributes<HTMLHeadingElement>
+>(({ className, ...props }, ref) => (
+    <h3
```
Let's use the `<Heading />` component.

adrians5j replied:
OK, although I've now noticed something else. But will post a comment when I'm done with changes.

leopuleo commented on packages/admin-ui/src/Card/Card.tsx:
```diff
@@ -1,56 +1,104 @@
     HTMLParagraphElement,
     React.HTMLAttributes<HTMLParagraphElement>
 >(({ className, ...props }, ref) => (
     <p ref={ref} className={cn("text-sm text-muted-foreground", className)} {...props} />
```
Let's use the `<Text>` component.

adrians5j replied:
OK, although I've now noticed something else. But will post a comment when I'm done with changes.

leopuleo commented on packages/admin-ui/src/Card/Card.tsx:
```diff
@@ -1,56 +1,104 @@
+
+const CardFooter = makeDecoratable("CardFooter", CardFooterBase);
+
+interface CardProps extends Omit<CardRootProps, "content"> {
```
I cannot find any reference to the close icon. Is it intentional?

![CleanShot 2024-09-02 at 11 25 11](https://github.com/user-attachments/assets/ff41ae82-59fc-471c-b629-024dd94d24e4)

adrians5j replied:
Added.

leopuleo commented on packages/ui/src/Elevation/Elevation.tsx:
```diff
@@ -19,15 +19,12 @@ export type ElevationProps = {
-    );
-});
+const Elevation = (props: ElevationProps) => {
+    return <Card {...props} content={<CardContent>{props.children}</CardContent>} />;
```
I would create an `Elevation` component in the new `admin-ui` package and reference it here and inside the `Card` component.

adrians5j replied:
This one I didn't quite understand 🤔 

What's interesting is that now there's an `elevation` prop on Card, so maybe we don't need a separate `Elevation` component?

leopuleo commented on packages/admin-ui/src/Card/Card.tsx:
```diff
@@ -1,56 +1,104 @@
+    headerDescription?: React.ReactElement<typeof CardDescriptionBase>;
+    content?: React.ReactElement<typeof CardContentBase>;
+    footer?: React.ReactElement<typeof CardFooterBase>;
+}
```
I don't see any references to some variants defined in the Figma file, such as: 

- Elevation
- Border Radius
- Paddings

Is it intentional?

adrians5j replied:
Added.

adrians5j committed:
wip: `header` prop

adrians5j committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/new-admin-ui' into feat/card

adrians5j committed:
wip: create Card component

adrians5j committed:
wip: create Card component

adrians5j commented:
Pushed another batch of changes.

Added all of the things you mentioned, thanks for the feedback! 
TBH I didn't even look much at the Figma file, hence all of the extra stuff missing, sry about that 🙃🫠

Besides adding the missing stuff, after playing with it more that I would've wanted to, I actually converted the `footer` and `content` props back into regular `React.ReactNode` props. In other words, users won't need to import `CardFooter` / `CardContent` components and use these upon defining card footer and content. In short, it was complicating some things. This way complication doesn't exist, plus it's a simpler API to use.

> [!NOTE]
> I've also taken into consideration the fact that, when using extra components for content/footer, users can pass extra props to these sections (like `data-testid`). I've also went with the assumption that this won't be that common of a use case for now. That was one additional vote for me going with the simpler API.

One other change I did here is that I introduced also the Icon component, which is now used for the close button in the card.

adrians5j committed:
wip: create Elevation component

adrians5j committed:
Merge branch 'refs/heads/feat/new-admin-ui' into feat/card

adrians5j committed:
wip: create Elevation component

adrians5j committed:
wip: create Elevation component

adrians5j committed:
wip: add color/size variations to Icon

adrians5j committed:
wip: add color/size variations to Icon

adrians5j committed:
wip: use components for `header` / `content` and `footer` props

adrians5j commented:
OK so another batch of updates here. Hoping we're finally ready to merge, but yeah, let's see what you have to say.

One final thing I did here (on top of everything we've talked about these days), is that, ultimately, I decided to go with the following props API:

```
    header: <CardHeader title="This is a card title" description="This is a card description" />,
    content: <CardContent content={"This is card content. Anything can go in here."} />,
    footer: <CardFooter content={"This is card footer. Anything can go in here."} />
```

So, no primitive values sent, just components (`CardHeader`, `CardContent`, `CardFooter`), which can actually take primitive values. 

Note that `title` and `description` props on `Header` component accept `React.ReactNode`, where, if a `string` was passed, then we use `Heading` / `Text` components internally to render texts. If not `string`, then we render just whatever the user has passed to it.

> [!NOTE]
> Have in mind that `Heading` / `Text` components only accept `string` via their `text` prop.

leopuleo reviewed (approved):
Adrian, thank you so much for all the hard work you put into this PR and the design around the APIs - it was definitely not an easy task! 😄

I think there are just two small things left to do:

- Remove @rmwc/elevation from @webiny/ui.
- Rewire the Icon component in the @webiny/ui package to the newly developed one.

These aren't blockers, but it would be great to address them before we merge.

Fantastic job, and thanks again! 🍻

adrians5j committed:
wip: finalize PR

adrians5j commented:
Hehe np Leo, glad we got this finally sorted!

And thank _you_ for all the reviewing you did here, awesome job as well!

Team work 💪

P.S. I resolved the two things you mentioned before merging.
